Unify C++ API with C++ extensions

goldsborough opened the pull request:
Currently the C++ API and C++ extensions are effectively two different, entirely orthogonal code paths. This PR unifies the C++ API with the C++ extension API by adding an element of Python binding support to the C++ API. This means the `torch/torch.h` included by C++ extensions, which currently routes to `torch/csrc/torch.h`, can now be rerouted to `torch/csrc/api/include/torch/torch.h` -- i.e. the main C++ API header. This header then includes Python binding support conditioned on a define (`TORCH_WITH_PYTHON_BINDINGS`), *which is only passed when building a C++ extension*.

Currently stacked on top of https://github.com/pytorch/pytorch/pull/11498

Why is this useful?

1. One less codepath. In particular, there has been trouble again and again due to the two `torch/torch.h` header files and ambiguity when both ended up in the include path. This is now fixed.
2. I have found that it is quite common to want to bind a C++ API module back into Python. This could be for simple experimentation, or to have your training loop in Python but your models in C++. This PR makes this easier by adding pybind11 support to the C++ API.
3. The C++ extension API simply becomes richer by gaining access to the C++ API headers.

@soumith @ezyang @apaszke

fmassa commented on torch/csrc/api/include/torch/python.h:
```diff
@@ -0,0 +1,106 @@
+py::class_<M, Extra...> add_module_bindings(py::class_<M, Extra...> module) {
+  return module.def("train", [](M& module) { module.train(); })
+      .def("eval", [](M& module) { module.eval(); })
+      .def("is_training", [](M& module) { return module.is_training(); })
```
Maybe on the python side you want to keep it the same, and make it a property that is called `training`?

goldsborough replied:
Sounds good

fmassa commented on torch/csrc/api/include/torch/python.h:
```diff
@@ -0,0 +1,106 @@
+      .def("is_training", [](M& module) { return module.is_training(); })
+      .def("zero_grad", [](M& module) { module.zero_grad(); })
+      .def("cuda", [](M& module) { module.to(torch::kCUDA); })
+      .def("cpu", [](M& module) { module.to(torch::kCPU); })
```
Could we also bind the `.to` method to python? This is the recommended way now I believe

goldsborough replied:
I have to check if a `torch.device` will convert into the right thing on Python side, otherwise yes good idea

fmassa reviewed (commented):
Nice!

I have a few questions regarding model serialization in Python.
It might be worth looking into also adding equivalents of `load_state_dict` and `state_dict` maybe? Also, I have the impression that a cpp module extension will not be seen by the python `nn.Module` as a `nn.Module`, so it might not recurse into it, is that right?

fmassa commented:
More generally, would it be possible (or worth it) making the cpp module extensions (when loaded in Python) inherit from `nn.Module`? Or would that require that `nn.Module` is moved entirely to Python?

goldsborough commented:
@fmassa That's a good and important question! The definite answer here, however, is that Python API and C++ API modules are currently not inter-operable at all, and so making one a submodule of the other will not work and is not intended to work right now. There could be future work of unifying them, but right now `nn::Module` and `nn.Module` are completely different things and cannot be mixed. In the future we may be able to give them a common C++ base class, along with JIT ScriptModule's too.

My intention here is simply to make it easier to add Python bindings to the C++ API, but not to make the C++ API inter-operate better with the Python API. The use case here is that you have a C++ API model and would like to do something with it in Python for experimentation. Another use case (which we have internally) is that you could keep two copies of your model, one in the Python API which you use for quick experiments, and one in the C++ API which you use in prod, and then have your training loop, data wrangling and visualization in Python, and instantiate either the C++ or Python model dynamically inside Python. The rest of the training code should work on either.

What this unification does give to C++ extension users is the utilities the C++ API provides, that is simple things like `torch::cuda::is_available()` or the `torch::NoGradGuard` (equivalent to `with torch.no_grad()`).

For us developers it is useful because we no longer have to think of the C++ extension codebase (currently torch/csrc/torch.h) and the C++ API codebase as separate.

fmassa commented:
@goldsborough I see, thanks for the explanation!
If we document exposing `torch::nn` into Python, it might be worth mentioning that we are not supposed to use the C++ Extension nn modules inside python nn modules then.

dzhulgakov commented on torch/csrc/api/include/torch/serialization.h:
```diff
@@ -2,45 +2,69 @@
   cereal::BinaryOutputArchive archive(stream);
   archive(*obj);
+#else
+  AT_ERROR("PyTorch compiled without serialization support");
```
I'd add here "without serialization support for C++ modules, it's WIP for the unified tensor format"

goldsborough commented:
@ezyang could you review this?

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -148,6 +148,7 @@ class BuildExtension(build_ext):
     def build_extensions(self):
         self._check_abi()
         for extension in self.extensions:
+            self._add_compile_flag(extension, '-DTORCH_WITH_PYTHON_BINDINGS')
```
I'm not a big fan of macros in this style, because they are "infectious": if I write a header that includes `torch.h` and I build my code with `-DTORCH_WITH_PYTHON_BINDINGS`, *everyone* downstream must also specify this macro.

I agree that having two header files named `torch/torch.h` is suboptimal, and agree it should be fixed. What I don't understand is why `torch/torch.h` has to either be Python-enabled or not, depending on context. Wouldn't it make more sense for there to be a Torch-with-Python header?

ezyang commented:
C++ looks ok, but I'm concerned about the macro.

goldsborough committed:
Unify C++ API and C++ extensions

More unification

Remove ../ include in one C++ extension

Solve the constructor problem for ModuleHolder

Create extension.h and deprecate torch/torch.h for C++ extensions

Include serialize/*.h in conda package

Skip cpp_api_extension test on Windows

Small improvements

goldsborough committed:
Add C++ frontend include to TorchConfig.cmake
